Invoices

abdellani opened the pull request:
#### What? Why?

- Closes # <!-- Insert issue number here. -->

related topics
https://github.com/openfoodfoundation/openfoodnetwork/issues/10724
https://github.com/openfoodfoundation/openfoodnetwork/issues/10725

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->



#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- enable invoices on the features toggle
- check the invoice on the order details page.
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/16113997/60b821b2-3b9e-4a64-aefa-6364054ea933)
- generate a new invoice.
- check the invoices list.
- try to update the order, then try to generate the invoice. Depending on the update (please check this [document](https://docs.google.com/spreadsheets/d/1hOM6UL4mWeRCFLcDQ3fTkbhbUQ2WvIUCCA1IerDBtUA/edit#gid=0)):
   - a new invoice will be generated.
   - the latest invoice will be updated.
#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category: User facing changes | Technical changes

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

abdellani commented:
Hi @openfoodfoundation/developers @openfoodfoundation/testers 

I wrote a draft of one possible way to implementation of the new requirements related to the invoices.

I added a new model `Invoice` 
* order_id: reference to the order
* status: can be `active` or `canceled` (from what I can see on https://github.com/openfoodfoundation/wishlist/issues/401) 
* Number: Number of the invoice ( I still don't know if it's just a number, or if it has a format for example 16-001)
* position: to sort the invoices (maybe we can use created_at instead ?)   
* data: will store the data used to render the invoice.

Questions: 
1. Is there something that needs to change in the model? 
2. Is it better to use `Invoice` or `Spree::Invoice` as a name for the model?  
`Spree::Admin::BaseController` expects the order's associated models to start with `Spree`   
https://github.com/openfoodfoundation/openfoodnetwork/blob/2daa49406abdae564308acdc13469abde6872824/app/controllers/spree/admin/base_controller.rb#L34-L39

I did a quick search for the code used for printing, I found that we trigger printing from 3 different parts (please correct me if I'm wrong):

1. `Spree::Admin::OrdersController#print`: 
    * Called when the user clicks `Print Invoice` on the Order's edit page. 
![image](https://user-images.githubusercontent.com/16113997/223965924-bbcbf58f-2761-4449-86dd-2af852f5831a.png)

2. `Spree::Admin::OrdersController#invoice` (relais on `OrderMailer`): 
     * Called when the user clicks `Send Invoice` on the Order's edit page.
3. `Spree::Admin::InvoicesController#create` (Bulk Invoice):  
    * Called when multiple orders are selected
![image](https://user-images.githubusercontent.com/16113997/223968109-e2a11c80-45da-4869-a4f0-e8ddd09e34e9.png)

This diagram when how classes' dependencies:
![image](https://user-images.githubusercontent.com/16113997/223967331-a3f01e49-cf8e-4155-a69e-3cb5cebd8eb0.png)

We'll need to add a new endpoint (or edit an existing one) to let the user generate a specific version of an invoice. 
![image](https://user-images.githubusercontent.com/16113997/223970269-cba1c49e-c544-480d-b6ce-97a6fbca5a6e.png)

Questions:

1. Is there a need to send old invoices by email?
2. what will be the default behavior if an order doesn't have any invoices? (bulk invoice, sending order)
3. In the case of the bulk invoice, do use the latest invoice version for every order? or do we give the possibility to select a specific version for some orders? 
4. what are the different invoice statuses? (from the screenshot: 'active', and 'canceled', but are there others?) 
5. How is the invoice status connected to the order status (and shipment, payment, etc...)? for example: if the order is canceled, will this move the invoice to `canceled` status?

RachL commented:
@abdellani thank you !! Here are some answer:

>  I still don't know if it's just a number, or if it has a format for example 16-001)

So far we chose to handle numbering in the future, so we would just keep order number for now (like we do today). Otherwise it legally needs to be sequential number per shop. 

This is the example I've put in the epic description:
>    JB orders at Filipe shop
    Filipe generates an invoice to JB. Let's say this invoice is number 30
    Rachel orders at Filipe. when Filipe invoice Rachel, this invoice has number 31
    Let's say JB now ask Filipe to make changes on the order (add a product for example)
    Filipe generates a new order for JB: this invoice has now the number 32 if the order was unpaid. If JB had already paid, two invoices are generated: one credit invoice with number 32 (with a mention that this credits invoice 30) and one new invoice with number 33.
Ideally, we would allow each hub to set an invoice prefix (in business details for example) and the number at which their invoices should start. From there, all invoices from this enterprise have the same prefix and a sequential number.

--

> Is there a need to send old invoices by email?

What do you mean by "old" invoices. The ones we have currently? As they are not stored this work will completely replace the old invoice format.

> what will be the default behavior if an order doesn't have any invoices?

Bulk invoice printing needs to generate the invoice if there isn't any. I would say same for bulk sending. However note the first PR should only focus on the edit order page. As we are working with a feature toggle we can work bits by bits.

> In the case of the bulk invoice, do use the latest invoice version for every order? or do we give the possibility to select a specific version for some orders?

When bulk invoice, we need to generate invoices, which means:
- if the order has changed, we will need to cancel the previous invoice and create a new one
- if the order has not changed, we generate an invoice if there wasn't any, or we use the one already there

> what are the different invoice statuses? 

A bit linked to the question I've asked here: https://github.com/openfoodfoundation/wishlist/issues/401#issuecomment-1443751391
If capturing a payment would pass the invoice in a "paid" status. I think we would have paid/unpaid/cancelled. Active would be replace by paid or unpaid. What do you think?

> How is the invoice status connected to the order status (and shipment, payment, etc...)? for example: if the order is canceled, will this move the invoice to canceled status?

Shipment is not displayed currently on the invoice I think, so we can leave it aside. If the order is canceled, it depends on the payment status. 
If the order is unpaid, the invoice is cancelled.
If the order was paid (partially or totally) - these are the test cases 8 and 9 I've tried to describe in 401, but perhaps I've forgotten something:

> 8. Generate a new invoice : if the order is unpaid, check the previous one is cancelled. If the order is paid, check there are now 3 lines in the table / 3 files available for download : 1. The previous invoice cancelled 2. The credit invoice 3. The new invoice with new amount

> 9. Credit invoices are issued when the full amount of the original order must be refunded. But there is also the case of the partial refund: an item gets removed for example and a refund must be done. In this case, the amount the order is reduced from must be displayed as line item and order total be re-calculated.

So credit invoices cannot be paid, they can only be refunded. Do we want to add a status for this?


This makes me realize that maybe just working on the edit page is too big. Maybe we should divide the work even more and work with one issue per action (delete product, capture payment, partial refunds...)?

mkllnk commented:
> position: to sort the invoices (maybe we can use created_at instead ?)

I would think that we don't need manual sorting. As you suggest, they should be sorted by creation date for now. Once we have sequential invoice numbers then we can order by those.

> data: will store the data used to render the invoice.

This is an interesting part. If we change the data format in the future, we probably need to run a migration for all old invoices. We'll see.

> Is it better to use Invoice or Spree::Invoice as a name for the model?

Just `Invoice`. The Spree namespace is a legacy which I would like to see removed one day.

mkllnk commented on db/migrate/20230308075421_create_invoices.rb:
```diff
@@ -0,0 +1,16 @@
+  def change
+    create_table :invoices do |t|
+      t.references :order, foreign_key: true, foreign_key: { to_table: :spree_orders }
+      t.string :status
+      t.integer :number
+      t.integer :position 
+      t.jsonb :data
+      t.datetime :date, default: -> { "CURRENT_TIMESTAMP" }
```
These should probably be `nil: false`.

mkllnk replied:
If it's just called `date`, should we just store a date then, without time?

mkllnk commented:
> Spree::Admin::BaseController expects the order's associated models to start with Spree

We have other models like Enterprise and Customer which are not in the Spree namespace if you are looking for some inspiration.

abdellani commented:
@RachL 

> What do you mean by "old" invoices. The ones we have currently? As they are not stored this work will completely replace the old invoice format.

No, I referred to:
![image](https://user-images.githubusercontent.com/16113997/224255517-945919bf-6e35-4f6a-b2e8-f35504e46ae0.png)

I don't think we need to send them, but I just wanted to confirm.

> When bulk invoice, we need to generate invoices, which means:
>
>if the order has changed, we will need to cancel the previous invoice and create a new one
>if the order has not changed, we generate an invoice if there wasn't any, or we use the one already there

In my opinion, we should have the same behavior when the user clicks on `Send invoice` or `Print invoice`.   What do you think?

![image](https://user-images.githubusercontent.com/16113997/224254357-600cd242-1564-4f01-a3c9-fdfd75c4708c.png)


> Bulk invoice printing needs to generate the invoice if there isn't any. I would say same for **`bulk sending`**. However note the first PR should only focus on the edit order page. As we are working with a feature toggle we can work bits by bits.

What do you mean by `bulk sending`? 

Do you mean sending multiple invoices at the same time? 
if yes, I missed that on the code. Can you please show me how can I trigger that from the UI?

![image](https://user-images.githubusercontent.com/16113997/224256216-0f9aab9e-4e93-4501-bfc8-710d47718631.png)

> If capturing a payment would pass the invoice in a "paid" status. I think we would have paid/unpaid/cancelled. Active would be replace by paid or unpaid. What do you think?
>...
>So credit invoices cannot be paid, they can only be refunded. Do we want to add a status for this?


With the current progress, I don't know yet. I think we can decide later.

abdellani commented:
@mkllnk 

For the first step, here's how I see the solution:

`InvoiceRenderer` will receive an `invoice` instance instead of an `order`.

The invoice template (`spree/admin/orders/invoice2`) will no longer have direct access to the order.

To generate the `data` attribute on the `Invoice`, I'll add a new class `InvoiceDataGenerate` to serialize the orders.

I suggest using `to_json` to serialize the orders, the result of their formatting methods (`display_item_total` for example), and their related associations (`payments` and `line_items`). 
For example:
```ruby
order.to_json(only: :note, methods: :display_item_total  ,include:[distributor:{ only: :name }])
```

To say if one invoice is no longer active, and we can generate a new invoice:
1. load the latest invoice of the order `latest_invoice`
2. serialize the current order `current_state_invoice`
3. Compare the two invoices json.
4. Some fields will be ignored in the comparison, for example `the enterprise legal name` 

From https://github.com/openfoodfoundation/wishlist/issues/401

> If the enterprise legal name changes or the address changes, currently we don't regenerate invoices, we keep previous invoices as is.


what do you think?

RachL commented:
@abdellani 

> I don't think we need to send them, but I just wanted to confirm.

Ah yes, correct!

> In my opinion, we should have the same behavior when the user clicks on Send invoice or Print invoice. What do you think?

Agreed.

> Do you mean sending multiple invoices at the same time?
if yes, I missed that on the code. Can you please show me how can I trigger that from the UI?

Sorry the PR is this in test ready: https://github.com/openfoodfoundation/openfoodnetwork/pull/10337

mkllnk commented:
> To generate the data attribute on the Invoice, I'll add a new class InvoiceDataGenerate to serialize the orders.

Good idea. But we could also create a data model inside the invoice class: `Invoice::Data`. The initializer takes a hash but it can also have a method to serialize an order: `Data.from_order(order)`

The advantage of this approach is that it contains all the knowledge of the data structure and it can implement equality checks. So if we override `hash` which is used for `eql?`  (and I think for `==`) then we can easily compare `current_data == stored_data`. What do you think?

I would create a hash for significant fields and use equality on those hashes.

abdellani commented:
@mkllnk 
 
After trying this approach of serializing the Order, I faced some issues with:

1. The models' methods, there are two types
    1. The methods that return simple values like `integers` and `booleans`. 
       For example: `order.outstanding_balance?` and `order.paid?`. Those are easy to serialize.
    2. The methods that return other models like: `order.sorted_line_items`.
2. Helpers for example:
    * `Order::Helper.last_payment_method`
    * `Spree::PaymentMethodsHelper.payment_method_name`
    * `OrderHelper.outstanding_balance_label`

For the models' methods:
In the beginning, I thought about serializing the output of the methods. The problem with this approach is that if any of the methods is updated to generate a different output, this will probably invalidate all the existing invoices.

Is it better to serialize the output of the models` methods? (and maybe avoid comparing the output of the methods during the comparison of the invoices?)
Or to write new methods that will use the serialized data? with this approach, we can edit the models without worrying about the invoices.

For the helpers:
the same problem and the same question.

mkllnk commented:
> For the models' methods:
> In the beginning, I thought about serializing the output of the methods. The problem with this approach is that if any of the methods is updated to generate a different output, this will probably invalidate all the existing invoices.

Updates will need care from now on. Any output can change. For example, a price can be `2.5` now but may become "$2.50" in the future. I reckon that we will have to migrate the data when that happens. So if we change any of the output for invoices, we have to migrate existing invoices to the new format. Alternatively, we could support different versions of the invoice format and have several templates but that would become a lot over time.

I'm wondering if we should use an `ActiveModel::Serializer` to represent all data needed for an invoice. We can call `to_json` on it to store the data. But reading a hash in the view could be awkward. So maybe we would have an `InvoiceDataPresenter` which takes the JSON and has some "helper" methods for the view. These separate helpers can also be useful to present data in different ways. So you have scope to compute the result from the data. You just have to be careful that you can't use any data from the environment, for example the default currency. All needed data has to be in the data object. So we probably want to avoid classes like Spree::Money in the presenter.

Does that help?

So store only needed data. Have a stable interface via a serializer (change of serializer requires a database migration). And create a presenter with new helper methods. Or can you think of a simpler way?

abdellani commented:
I like the idea of using `ActiveModel::Serializer`

Before your comment, I tried to serialize using `to_json()`. This approach will generate a huge JSON object that will be difficult to manipulate. I felt that it'll be difficult to maintain.
```ruby
    order.to_json({
                    only: [:number, :special_instructions, :note],
                    methods: [:outstanding_balance?, :display_outstanding_balance, :paid?],
                    include: {
                      distributor: {
                        only: [:name, :abn, :invoice_text],
                        include: {
                          address: {
                            methods: [:full_address]
                          },
                          business_address: {
                            only: [:company]
                          },
                          contact: {
                            only: [:email]
                          }
                        }
                      },
                      bill_address: {
                        only: [:phone],
                        methods: [:full_address]
                      },
                      customer: {
                        only: [:code, :email]
                      },
                      order_cyle: {
                        only: [:name]
                      },
                      shipping_method: {
                        only: [:require_ship_address]
                      },
                      ship_address: {
                        only: [:phone],
                        method: [:full_name]
                      },
                      payments: {
                        only: [:name, :state, :description, :created_at],
                        #missing method name
                        methods: [:display_amount]
                      },
                      line_items: {
                        only: [],
                        include: {
                          supplier: {
                            only: [:name]
                          }
                        }
                      }
                    }
                  })
```

abdellani commented:
@mkllnk 

I started by adding new serializers (`app/serializers/invoice`) that will be used on the invoices.
I updated the template to use the invoice's data presenter. The presenter is composed of objects that match the data model.

I guess I'm on the right path :)

what do you think?

mkllnk commented on app/models/spree/order.rb:
```diff
@@ -615,6 +616,20 @@ def sorted_line_items
+    def can_generate_new_invoice?
+      return true if invoices.empty?
+
+      invoices.last.data["order"].to_json != to_json(include: :line_items)
```
This check needs changing now, right? Or is `to_json` finding the right serializer already?

abdellani replied:
I extracted the comparison logic into a separated class.
app/services/order_invoice_comparator.rb

I use a non-recursive BFS to compare invoices.

mkllnk commented on app/models/spree/order.rb:
```diff
@@ -615,6 +616,20 @@ def sorted_line_items
+    end
+
+    def generate_invoice(number = invoices.size + 1, date = Time.now)
+      invoices.create({
+                        number: number,
+                        date: date,
+                        data: to_json(include: :line_items)
+                      })
```
Do you need the hash brackets here?

mkllnk commented on app/models/spree/order.rb:
```diff
@@ -615,6 +616,20 @@ def sorted_line_items
+      invoices.create({
+                        number: number,
+                        date: date,
+                        data: to_json(include: :line_items)
```
Is `to_json` already creating a string? I thought the model was expecting a hash and it's serialized when stored to the database.

mkllnk commented on app/views/spree/admin/invoices/_invoices_table.html.haml:
```diff
@@ -0,0 +1,24 @@
+      %th= t(:file)
+  %tbody
+    - @order.invoices.each do |invoice|
+      - tr_class = cycle('odd', 'even')
```
Oh dear, this is really old code. CSS is much better at the odd-even style. But that doesn't need solving in this PR.

mkllnk reviewed (commented):
Yes, you are on a good path. I think it's time to squash some of the fixes and add some specs.

abdellani commented:
@mkllnk 

I added some changes:

* refactored the presenters.
* Implemented an invoice comparator

Can you please review those changes?

mkllnk commented on app/services/order_invoice_comparator.rb:
```diff
@@ -0,0 +1,46 @@
+      queue1 = [tree1]
+      queue2 = [tree2]
+
+      while  !queue1.empty? && !queue2.empty?
```
This means if one queue has additional objects but they are all `nil` then we compare them with the `nil` result of `shift`ing an empty array. Those trees are seen as equal. Is that intended?

abdellani replied:
This will disappear on the recursive version.

abdellani commented on app/services/create_invoice.rb:
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+# Create an invoice for the order
+class CreateInvoice
```
Yes, we don't need this class.
When I created this class it was doing more. I'll remove it now.

abdellani commented on spec/factories/invoice_factory.rb:
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :invoice, class: Invoice do
+    after(:create) do |invoice|
+      invoice.update!(data: invoice.order.serialize_for_invoice)
```
Yeah, we can use `after_create` hook

abdellani commented on app/services/order_invoice_comparator.rb:
```diff
@@ -0,0 +1,46 @@
+class OrderInvoiceComparator
+  def equal?(invoice1,invoice2)
+      # We'll use non-recursive version of the BFS algorithm to find if current state of invoice is outdated
+      # the root will be the order
+      # On each node, we'll a list of relevant attributes that will be used on the comparison
```
I think we can use a recursive version.

mkllnk commented on app/controllers/spree/admin/invoices_controller.rb:
```diff
@@ -6,13 +6,34 @@ class InvoicesController < Spree::Admin::BaseController
+      def generate
+        @order = Order.find_by_number(params[:order_id])
+        if @order.can_generate_new_invoice?
+          @order.invoices.create!({
+                                    date: Time.now.to_date,
+                                    number: @order.next_invoice_number,
+                                    data: invoice_data
+                                  })
+        elsif @order.can_update_latest_invoice?
+          @order.invoices.last.update!({
+                                          date: Time.now.to_date,
+                                          data: invoice_data
+                                        })
```
Since another PR is just working on improving layout and line length, I would try to reduce the indent which also means that you don't have to change it when the method call name is changed. And it's also recommended to always use `Time.zone`. And there's a method `today`:


```suggestion
          @order.invoices.create!(
            date: Time.zone.today,
            number: @order.next_invoice_number,
            data: invoice_data
          )
        elsif @order.can_update_latest_invoice?
          @order.invoices.last.update!(
            date: Time.zone.today,
            data: invoice_data
          )
```

mkllnk commented on app/services/order_invoice_comparator.rb:
```diff
@@ -0,0 +1,54 @@
+    bfs(current_state_invoice.presenter, latest_invoice.presenter, invoice_update_selector)
+  end
+
+  def bfs(node1, node2, attributes_selector)
```
`bfs` is the type of algorithm but it would be nice if the name told us what the return value will be. I would suggest:
```suggestion
  def different?(node1, node2, attributes_selector)
```

mkllnk commented on app/services/order_invoice_comparator.rb:
```diff
@@ -0,0 +1,54 @@
+
+    return true if presenters1.size != presenters2.size
+
+    presenters1.zip(presenters2).each do |presenter1, presenter2|
+      return true if bfs(presenter1, presenter2,attributes_selector)
+    end
+    false
```
Good idea with `zip`. In conjunction with my naming suggestion above, I would write this with `any?`:
```suggestion
    presenters1.zip(presenters2).any? do |presenter1, presenter2|
      different?(presenter1, presenter2, attributes_selector)
    end
```

mkllnk reviewed (commented):
Great. I'm looking forward to the final PR.

abdellani commented on app/controllers/spree/admin/base_controller.rb:
```diff
@@ -33,7 +33,7 @@ def warn_invalid_order_cycles
       def model_class
         const_name = controller_name.classify
-        return "Spree::#{const_name}".constantize if Spree.const_defined?(const_name)
+        return Spree.const_get(const_name) if Spree.const_defined?(const_name)
```
```diff
-        return "Spree::#{const_name}".constantize if Spree.const_defined?(const_name)
+        return Spree.const_get(const_name) if Spree.const_defined?(const_name)
```
A problem occurs when `const_name` = `Invoice`.

 the requests don't hit the right controller `Spree::Admin::InvoicesController`.
I'm not sure about the source of the problem, but here's the current behavior:

```ruby
3.0.3 :018 > Object.const_defined?("Spree::Invoice")
 => false 
3.0.3 :019 > Spree.const_defined?("Invoice") # From this, I expect Spree::Invoice to be defined
 => true 
3.0.3 :020 > "Spree::Invoice".constantize # < I can't explain this !
/.../activesupport-7.0.4.3/lib/active_support/inflector/methods.rb:280:in `const_get': uninitialized constant Spree::Invoice (NameError) 
```
I think it's better to use `Object.const_defined?`
```diff
-        return "Spree::#{const_name}".constantize if Spree.const_defined?(const_name)
+        return "Spree::#{const_name}".constantize if Object.const_defined?("Spree::#{const_name}")
```

abdellani replied:
I found this topic  but I'm not it helps
https://community.openfoodnetwork.org/t/controllers-structure-a-cleanup-strategy/1994/4

Matt-Yorkley commented on app/models/spree/order.rb:
```diff
@@ -618,6 +619,39 @@ def sorted_line_items
+    def can_generate_new_invoice?
+      return true if invoices.empty?
+
+      invoice_comparator.can_generate_new_invoice? current_state_invoice, invoices.last
```
Epic work! :heart:

I had a little thought here about maybe reshaping the boundaries and responsibilities between the `Spree::Order` and `OrderInvoiceComparator` classes and modifying their public interface a bit.

So I think if it was me I would move all these new methods/logic out of `Spree::Order` and into `OrderInvoiceComparator` so that the `Spree::Order` class isn't holding the responsibility for knowing about the details of the invoice generating process. For example, the `#can_generate_invoice?` method in `OrderInvoiceComparator` could take an `order` as an argument, and then all the methods here like  `#current_state_invoice` and `#serialize_for_invoice` could be private methods inside `OrderInvoiceComparator`.

Then elsewhere where these methods need to be called like in the invoices controller or the order helper, instead of calling `order.can_generate_invoice?` it would be `OrderInvoiceComparator.new.can_generate_invoice?(order)` or the order could be an argument in the initializer, like `OrderInvoiceComparator.new(order).can_generate_invoice?`.

That class already has the responsibility of figuring out the answers to these questions, right?

It's not necessary right now, but it might be a nice refactor at some point. What do you think? :thinking:

Matt-Yorkley replied:
Something along these lines...

```ruby
class OrderInvoiceComparator
  def initializer(order)
    @order = order
  end

  def can_generate_new_invoice?
    return true if @order.invoices.empty?
    
    different?(current_state_invoice.presenter, latest_invoice.presenter,
               invoice_generation_selector)
  end
  
  ...
  
  private
  
  def current_state_invoice
      Invoice.new(
        order: @order,
        data: serialize_for_invoice,
        date: Time.zone.today,
        number: @order.invoices.count + 1
      )
  end
  
  def latest_invoice
    @order.invoices.last
  end
  
  def serialize_for_invoice
    Invoice::OrderSerializer.new(@order).serializable_hash
  end
  
  ...
  
```

Hmmm and actually some of these methods could be memoized so they don't do the same work twice in the same thread if both `#can_generate_invoice?` and `#can_update_latest_invoice?` get called on the same order

```ruby
  def serialize_for_invoice
    @serialize_for_invoice ||= Invoice::OrderSerializer.new(@order).serializable_hash
  end
```
```ruby
comparator = OrderInvoiceComparator.new(order)
comparator.can_generate_invoice?
comparator.can_update_latest_invoice?
```

:thinking::thinking::thinking:

Anyway, feel free to ignore my suggestions if you don't want to add any more to this PR (it's big already)!

abdellani replied:
I like the idea, I'll update the code and request another review :)

Matt-Yorkley commented on app/models/invoice.rb:
```diff
@@ -0,0 +1,15 @@
+  end
+
+  def serialize_order
+    self.data ||= order.serialize_for_invoice
```
That method is removed from `Spree::Order` now, so:

```suggestion
    self.data ||= Invoice::OrderSerializer.new(order).serializable_hash
```

Matt-Yorkley commented on app/views/spree/admin/orders/_invoice_table4.html.haml:
```diff
@@ -0,0 +1,62 @@
+        %th{:align => "right", :width => "15%"}
+          %h5= t(:invoice_column_tax_rate)
+  %tbody
+    - @order.sorted_line_items.each do |item|
```
There's a few places in this `invoice_table4` file that are using `@order` instead of `@invoice_presenter`. I think they all need to be `@invoice_presenter`, otherwise the data from the invoice isn't being used.

Matt-Yorkley commented on app/views/spree/admin/orders/_invoice_table4.html.haml:
```diff
@@ -0,0 +1,62 @@
+    - @order.sorted_line_items.each do |item|
+      %tr
+        %td
+          = render 'spree/shared/line_item_name', line_item: item
```
I think you might need to use the `_invoice/line_item_name` partial here instead? 
```ruby
= render 'spree/admin/orders/_invoice/line_item_name', line_item: item
```

Matt-Yorkley commented on app/models/invoice/data_presenter.rb:
```diff
@@ -0,0 +1,114 @@
+
+    attributes :additional_tax_total, :currency, :included_tax_total, :payment_total,
+               :shipping_method_id, :state, :total
+    attributes :number, :note, :special_instructions, prefix: :order
```
Is there a reason for prefixing these three order attributes? 

I think it would be better if a regular `Spree::Order` object and an instance of a `Invoice::DataPresenter` data object have the same attribute names. That way they could be used interchangeably in more situations, like [Duck Types](https://en.wikipedia.org/wiki/Duck_typing).

```ruby
real_order_object = Spree::Order.find(23)
order_from_invoice = Invoice::DataPresenter.new(order.invoices.first)

real_order_object.number # => 23
order_from_invoice.number # => 23
```

Currently `order_from_invoice.number` would throw a `MethodMissing` error but `order_from_invoice.order_number` would return `23`. Know what I mean? I think there will be benefits later on if they both have the same interface.

So... that would mean removing the `prefix: :order` for these three definitions and, and then changing all instances here of `.order_number`, `.order_note`, and `.order_special_instructions` to `.note`, `.number`, and `.special_instructions`. And then if you've got invoices saved in your development database you would have to delete them with `Invoice.destroy_all` and remake them because the format will be changed.

One benefit of this change is that we would be able to render an invoice object completely independently of the `order` object, like:
```ruby
@order = Invoice.find(params[:invoice_id]).presenter

InvoiceRenderer.new.args(@order)
```

What do you think?

abdellani replied:
That was helpful.
Thank you @Matt-Yorkley  :pray:

mkllnk commented on spec/models/invoice_spec.rb:
```diff
@@ -0,0 +1,7 @@
+
+require 'spec_helper'
+
+RSpec.describe Invoice, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
This should be filled in.

mkllnk reviewed (changes requested):
Epic work. I still have some comments but they are minor. It would be great to get someone else to review. I find it very difficult to review something that I mostly reviewed already but not completely.

abdellani commented on db/schema.rb:
```diff
@@ -1192,7 +1203,7 @@
   create_table "vouchers", force: :cascade do |t|
     t.string "code", limit: 255, null: false
-    t.datetime "expiry_date"
+    t.datetime "expiry_date", precision: nil
```
fixed

Matt-Yorkley reviewed (approved):
Great, I think its ready for testing :+1:

mkllnk commented on app/models/invoice.rb:
```diff
@@ -0,0 +1,17 @@
+  end
+
+  def serialize_order
+    return data unless data.empty?
+
+    self.data = Invoice::OrderSerializer.new(order).serializable_hash
```
Just curious, when is the data not nil but empty?

abdellani replied:
@mkllnk 
```ruby
3.0.3 :003 > {}.nil?
 => false 
3.0.3 :004 > {}.empty?
 => true 
```

mkllnk replied:
Oh, of course, it's a hash.

mkllnk commented on lib/open_food_network/feature_toggle.rb:
```diff
@@ -38,6 +38,12 @@ module FeatureToggle
       "vouchers" => <<~DESC,
         Add voucher functionality. Voucher can be managed via Enterprise settings.
       DESC
+      "white_label" => <<~DESC,
+        Customize shopfront (shop, cart, checkout) and emails without OFN branding.
+      DESC
```
You accidentally added the white label feature again. It was removed.
```suggestion
```

dacook commented:
Great, is there sufficient information for testing? (I notice the description hasn't been updated since it was a draft)

abdellani commented:
> Great, is there sufficient information for testing? (I notice the description hasn't been updated since it was a draft)

I update the description, I totally forgot about it.
Thank you @dacook :+1:

sigmundpetersen commented:
Since this is already a large PR and it's behind toggle, maybe we could merge and create smaller issues for the points you found @drummer83 ? What do you think?

drummer83 commented:
Sure, no strong opinion here.
But I don't have time to do it today.

abdellani commented:
I shared the same opinion with @sigmundpetersen  :+1: , The PR is already big.
I'll be waiting for your comment @drummer83 
Thank you :+1:

drummer83 commented:
@abdellani 
I rebased and merged this PR. Thanks a lot!

I created issues for my findings and added them to the Invoices column of the board and to the Epic.

Let me know if you have any questions around this.
Thanks!
